### PR TITLE
Remove DirtyHandler and replace it with Prompt

### PR DESCRIPTION
**Pull request by nsams**
This refactor simplifies things a lot.

Also fix the long standing issue with subroutes triggering dirty handler by allowing any subroutes.

**We have one remaining issue:**

Consider the following example:
```
            <Switch>
                <Route path="/" exact={true}>
                    <RouterPrompt
                        message={() => {
                            return "sure?";
                        }}
                    />
                    <Link to="/bar">barLink</Link>
                </Route>
                <Route path="/bar">bar</Route>
            </Switch>
```
The second route (/bar) is not a subroute of the first route (/) and klicking barLink should trigger Dirty dialog. See also this unit test:
https://github.com/vivid-planet/comet/pull/1027/files#diff-7ab86bcbbf4488d6f29aa7536622d92292c13c2a8eacc8d5cba5b3ffd1f16c4dR106-R116

This is not a theoretical issue but happens for every Tabs or Stack component when the first (=initial) page contains a form.

Also possible to reproduce here:

http://localhost:26638/?path=/story/comet-admin-form--forminstack&knob-Locale=en&knob-Show%20Error%20Strategy=while-typing&knob-Theme=Comet
http://localhost:26638/?path=/story/comet-admin-form--form-with-router-tabs&knob-Locale=en&knob-Show%20Error%20Strategy=while-typing&knob-Theme=Comet
and this test https://github.com/vivid-planet/comet/pull/1027/files#diff-cf3e43fe3b6625879fd3ede1accc3368ae122d5c01e678d9dc6fd4ec13bbb603R66

Possible Solutions:

- Make every Prompt create a context (must nest now) and register all routes below the Prompt. With that information we know what routes are below a Prompt. Prototype: https://github.com/vivid-planet/comet/commit/0782061bfe00f5484e90e78e9e21d2a889ca0455
    - this adds quite some complexity
    - This makes a few incompatbilities with react-router:
        - Needs a Custom Router component
        - Switch doesn't work anymore (as it does render only one Route and so our system doesn't know about the not-rendered-ones)
        - Route always needs path to register correctly
- Avoid such routes by giving the initial StackPage/Tab a named route and redirect to that route
    - that causes issues with multiple stacks/tabs on one page <s>(which is currently not supported correctly anyway, but planned)</s>
- Make <RouterPrompt wrap the component containing the (allowed) subroutes and create an additional wrapping route that adds a sub-path to the subroutes
    - that requires a unique path, also for eg. every Form
    - implemented in https://github.com/vivid-planet/comet/pull/1052

### CODE REVIEW

File: packages/admin/admin-stories/src/admin/form/RouterTabs.tsx [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+                                    </Toolbar>
+                                    <RouterTabs>
+                                        <RouterTab label="Form 1" path="" forceRender={true}>
+                                            <Card variant="outlined">
+                                                <CardContent>
+                                                    <Field label="Foo" name="foo" component={FinalFormInput} />
+                                                </CardContent>
+                                            </Card>
+                                        </RouterTab>
+                                        <RouterTab label="Form 2" path="/form2" forceRender={true}>
```
Comment (nsams) [inline packages/admin/admin-stories/src/admin/form/RouterTabs.tsx:85]: this is now working correctly :)

File: packages/admin/admin/src/FinalForm.tsx [hunk 2]
```diff
@@ -35,8 +36,7 @@ interface IProps<FormValues = AnyObject> extends FormProps<FormValues> {
 }
 
 export function FinalForm<FormValues = AnyObject>(props: IProps<FormValues>) {
-    const client = useApolloClient();
-    const dirtyHandler = React.useContext(DirtyHandlerApiContext);
+    const { client } = React.useContext(getApolloContext());
```
Comment (nsams) [inline packages/admin/admin/src/FinalForm.tsx:39]: unrelated change to have a soft runtime dependency on apollo (client can be undefined now)

(that simplified story)

File: packages/admin/admin/src/router/PromptHandler.tsx [hunk 3]
```diff
@@ -5,13 +5,94 @@ import { Prompt } from "react-router";
...
+                    message,
+                    callback,
+                });
+            },
+        };
+
+    const promptMessage = (location: History.Location, action: History.Action): boolean | string => {
+        let ret: boolean | string = true;
+        Object.keys(registeredMessages.current).forEach((id) => {
+            if (location.pathname.startsWith(registeredMessages.current[id].path)) {
```
Comment (nsams) [inline packages/admin/admin/src/router/PromptHandler.tsx:44]: This is where navigating to subroutes gets allowed.

In theory there could be subroutes that replace the current component and prompt is not shown. Such routes have to be avoided.

File: packages/admin/admin/src/router/PromptHandler.tsx [hunk 4]
```diff
@@ -5,13 +5,94 @@ import { Prompt } from "react-router";
...
+                // allow transition if location is below path where prompt was rendered
+                return true;
```
Comment (johnnyomair) [inline packages/admin/admin/src/router/PromptHandler.tsx:46]: This only returns in `Object.keys(registeredMessages.current).forEach(...)`, is this intentional? If not, you'll need to use `for .. of`

File: packages/admin/admin/src/router/PromptHandler.tsx [hunk 5]
```diff
@@ -5,13 +5,94 @@ import { Prompt } from "react-router";
...
+
+    return (
+        <>
+            <RouterConfirmationDialog
+                isOpen={state.showConfirmationDialog}
+                message={state.message}
+                handleClose={handleClose}
+                showSaveButton={Object.keys(saveActions.current).length > 0}
+            />
+            <Prompt when={true} message={promptMessage} />
```
Comment (johnnyomair) [inline packages/admin/admin/src/router/PromptHandler.tsx:83]: Why don't you do the check here and use the `when` prop instead of returning true in `promptMessage`? Is this not possible?

File: packages/admin/admin/src/EditDialog.tsx [hunk 6]
```diff
@@ -63,17 +62,33 @@ export function useEditDialog(): [React.ComponentType<IProps>, { id?: string; mo
...
+    const saveActionRef = React.useRef<SaveAction>();
+
     const EditDialogWithHookProps = React.useMemo(() => {
-        return (props: IProps) => {
+        return (props: EditDialogProps) => {
             return (
                 <Selection>
                     <EditDialogFormApiProvider>
-                        <EditDialogInner {...props} selection={selection} selectionApi={selectionApi} api={api} />
+                        <RouterContext.Provider
```
Comment (johnnyomair) [inline packages/admin/admin/src/EditDialog.tsx:73]: The only reason for this additional RouterContext is to store the saveAction for the inner component, am I right? Isn't there a better way?

File: packages/admin/admin-stories/src/admin/form/RouterTabs.tsx [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+                    </Card>
+                </RouterTab>
+            </RouterTabs>
+        </Stack>
+    );
+}
+
+storiesOf("@comet/admin/form", module)
+    .addDecorator(storyRouterDecorator())
+    .add("RouterTabs with forms", () => <Story />);
```
Comment (johnnyomair) [inline packages/admin/admin-stories/src/admin/form/RouterTabs.tsx:110]:
```suggestion
    .add("Form with router tabs", () => <Story />);
```

File: packages/admin/admin/src/router/PromptHandler.tsx (hunk 5 above)
Comment (nsams) [reply packages/admin/admin/src/router/PromptHandler.tsx:83]: what check?

File: packages/admin/admin/src/router/PromptHandler.tsx (hunk 4 above)
Comment (nsams) [reply packages/admin/admin/src/router/PromptHandler.tsx:46]: hm, yes.... but how is this working correclty??

File: packages/admin/admin/src/EditDialog.tsx (hunk 6 above)
Comment (nsams) [reply packages/admin/admin/src/EditDialog.tsx:73]: Yes - I consider this a pretter clever mechanismn :)

File: packages/admin/admin/src/router/PromptHandler.tsx (hunk 5 above)
Comment (johnnyomair) [reply packages/admin/admin/src/router/PromptHandler.tsx:83]: The check to show the prompt

image


https://v5.reactrouter.com/core/api/Prompt

File: packages/admin/admin/src/router/PromptHandler.tsx (hunk 4 above)
Comment (johnnyomair) [reply packages/admin/admin/src/router/PromptHandler.tsx:46]: because `ret` in the outer scope is true by default, and you only change it below this return. The return in the forEach isn't used for anything

File: packages/admin/admin/src/EditDialog.tsx (hunk 6 above)
Comment (johnnyomair) [reply packages/admin/admin/src/EditDialog.tsx:73]: But why can't the inner component access the save action in the outer context directly?

File: packages/admin/admin/src/router/PromptHandler.tsx (hunk 5 above)
Comment (nsams) [reply packages/admin/admin/src/router/PromptHandler.tsx:81]: because message is called on demand (when location changes) and this allows us to check registeredMessages (which are non reactive!) on demand

File: packages/admin/admin/src/EditDialog.tsx (hunk 6 above)
Comment (nsams) [reply packages/admin/admin/src/EditDialog.tsx]: I moved it into EditDialogInner, no need to do this from the outer component

**Review by johnnyomair — DISMISSED** [review]

**Comment by dkarnutsch** [comment]
From JFX: TBC: are tests bundled with the package?

**Review by johnnyomair — APPROVED** [review]
